doc/ref/internals/fileserver-and-client.rst

```diff
@@ -0,0 +1,131 @@
+.. _internals-fileserver-client:
+
+The Salt Fileserver and Client
+==============================
+
+Introduction
+------------
+
+Salt has a modular fileserver, and mulitple client classes which are used to
+interact with it. This page serves as a developer's reference, to help explain
+how the fileserver and clients both work.
+
+Fileserver
+----------
+
+The fileserver is not a daemon, so the fileserver and client are not a true
+server and client in the traditional sense. Instead, the fileserver is simply a
+class (``salt.fileserver.Fileserver``), located in
+`salt/fileserver/__init__.py`_. This class has access to the configured
+fileserver backends via a loader instance, referenced as ``self.servers``. When
+a request comes in from the fileclient, it will ultimately result in a
+``Fileserver`` class function being run.
+
+The functions in this class will run corresponding functions in the configured
+fileserver backends to perform the requested action. So, in summary:
+
+1. A fileclient class makes a request...
+2. which triggers the fileserver to run a function...
+3. which runs a named function in each of the configured backends.
+
+Not all of the functions will always execute on every configured backend. For
+instance, the ``find_file`` function in the fileserver will stop when it finds
+a match, so if it finds a match for the desired path in the first configured
+backend, it won't proceed and try to find the file in the next backend in the
+list.
+
+Additionally, not all backends implement all functions in the
+``salt.fileserver.Fileserver`` class. For instance, there is a function called
+``update``, which exists to update remote fileservers such as the ``git``,
+``hg``, and ``svn`` backends. This action has no use however in the ``roots``
+backend, so it is simply not implemented there, and thus the ``roots`` backend
+will be skipped if the ``update`` function is run on the fileserver.
+
+Backends for the fileserver are located in `salt/fileserver/`_ (the files not
+named ``__init__.py``).
+
+.. _`salt/fileserver/__init__.py`: https://github.com/saltstack/salt/tree/develop/salt/fileserver/__init__.py
+.. _`salt/fileserver/`: https://github.com/saltstack/salt/tree/develop/salt/fileserver
+
+Fileclient
+----------
+
+There are three fileclient classes:
+
+salt.fileclient.RemoteClient
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This client is used when :conf_minion:`file_client` is set to ``remote``. This
+is how minions request files from the master.
+
+Functions in this client will craft a payload and send it to the master via the
+transport channel. This is the same way that the minion asks the minion to do
```

Comment:
`minion asks the minion`? I think this should be `minion asks the master` right?